File: scilab/modules/signal_processing/sci_gateway/fortran/sci_corr.f
c Scilab ( http://www.scilab.org/ ) - This file is part of Scilab
c Copyright (C) INRIA
c 
c This file must be used under the terms of the CeCILL.
c This source file is licensed as described in the file COPYING, which
c you should have received as part of this distribution.  The terms
c are also available at    
c http://www.cecill.info/licences/Licence_CeCILL_V2-en.txt
c     =======================================
      subroutine scicorr(fname)
c     implicit undefined (a-z)
      include 'stack.h'
      character*(*) fname
      logical checkrhs,getrmat,getscalar
      logical cremat,getexternal,getmat
      integer m1,n1,lr1,m2,n2,lr2,lr,top2, iadr,sadr
      integer topk,topw,gettype
      double precision yyy
c
      logical fort,type
      integer fini,update
c
      external dgetx,dgety,setdgetx,setdgety
      external bgetx,bgety
      character*(nlgh+1) namex,namey
      common / corrname / namex,namey
      common / corradr  / kgxtop,kgytop,ksec,kisc
      common / corrtyp /  itxcorr,itycorr
      common/  iercorr /iero
c
      data fini/15/,update/30/
      iadr(l)=l+l-1
      sadr(l)=(l/2)+1
c     calcul de correlations
      iero=0
      topk=top
      topw=top+1
      top2=top-rhs+1
      il1=iadr(lstk(top2))
      if(istk(il1).eq.10) then
         ich=il1+5+istk(il1+1)*istk(il1+2)
         ich=istk(ich)
         if(ich.eq.fini) goto 41
         if(ich.eq.update) goto 42
      endif
C     usual case [cov,mean]=corr(x,[y],nlags) 
C     ------------------------------------------
      if (.not.checkrhs(fname,2,3)) return
C     getting nlags (last argument)
      if(.not.getscalar(fname,topk,top,lr))return
      lag=int(stk(lr))
      if (.not.cremat(fname,topk+1,0,1,lag,lcxy,lc)) return
      if (.not.cremat(fname,topk+2,0,1,2,lrxy,lc)) return

      if(rhs.eq.3) then
         top=top-1
         if (.not.getrmat(fname,topk,top,m2,n2,lr2)) return
         ny= m2*n2
         top=top-1
         if (.not.getrmat(fname,topk,top,m1,n1,lr1)) return
         nx= m1*n1
         if(nx.ne.ny) then
            call error(60)
            return
         endif
      endif
      if(rhs.eq.2) then
         top=top-1
         if (.not.getrmat(fname,topk,top,m1,n1,lr1)) return
         nx= m1*n1
         lr2=lr1
      endif
      call tscccf(stk(lr1),stk(lr2),nx,stk(lcxy),stk(lrxy),
     $     lag,ierr)
      if(ierr.eq.-1) then
         buf='too many coefficients are required'
         call error(999)
         return
      endif
      top=topk-rhs+1
      call copyobj(fname,topk+1,top)
      if(lhs.eq.2) then
         top=top+1
         if (.not.cremat(fname,top,0,1,rhs-1,lr,lc)) return
         call unsfdcopy(rhs-1,stk(lrxy),1,stk(lr),1)
      endif
      return
C     ----------------------
C     Special case using fft 
C       [cov,mean]=corr('fft',xmacro,[ymacro],n,sect)
C     ----------------------
 41   continue
c     methode de la fft
      if (.not.checkrhs(fname,4,5)) return
c     -------last argument (sect)
      if(.not.getscalar(fname,topk,top,lr))return
      lag=int(stk(lr))
      mm=2*lag
c     --------total size 
      kntop=top-1
      if(.not.getscalar(fname,topk,kntop,lr))return
      n=int(stk(lr))
c     -------external for xmacro 
      kgxtop= top-rhs+2 
      itxcorr=gettype(kgxtop)
      if (.not.getexternal(fname,topk,kgxtop,namex,fort,
     $     setdgetx)) return
      mode=2
c     -------external for ymacro 
      if(rhs.eq.5) then
         kgytop= top-rhs+3
         itycorr=gettype(kgytop)
         if (.not.getexternal(fname,topk,kgytop,namey,type,
     $     setdgety)) return
         if( type.neqv.fort) then
            err=3
            call error(55)
            return
         endif
         mode=3
      endif
c     Working areas 
      if (.not.cremat(fname,top+1,0,1,mm,lxa,lc)) return
      if (.not.cremat(fname,top+2,0,1,mm,lxr,lc)) return
      if (.not.cremat(fname,top+3,0,1,mm,lxi,lc)) return
      if (.not.cremat(fname,top+4,0,1,sadr(mm),lzr,lc)) return
      if (.not.cremat(fname,top+5,0,1,sadr(mm),lzi,lc)) return
      top=top+6
      ksec=top
      if (.not.cremat(fname,ksec,0,1,1,lr,lc))return
      kisc=top+1
      if (.not.cremat(fname,kisc,0,1,1,lr,lc))return
      if(fort) then
         call cmpse2(mm,n,mode,dgetx,dgety,stk(lxa),stk(lxr),
     1        stk(lxi),stk(lzr),stk(lzi),ierr)
         if(ierr.gt.0) then
            buf='fft call : needs power of two!'
            call error(999)
            return
         endif
      else
         call cmpse2(mm,n,mode,bgetx,bgety,stk(lxa),stk(lxr),
     1        stk(lxi),stk(lzr),stk(lzi),ierr)
         if(err.gt.0.or.err1.gt.0) return
         if(ierr.gt.0) then
            buf='fft call : needs a power of 2'
            call error(999)
            return
         endif
      endif
      top=topk-rhs+1
      if (.not.cremat(fname,top,0,1,lag,lr,lc)) return
      call unsfdcopy(lag,stk(lxa),1,stk(lr),1)
      if(lhs.eq.1) return
      top=top+1
      if (.not.cremat(fname,top,0,1,rhs-3,lr,lc)) return
      call unsfdcopy(rhs-3,stk(lxr),1,stk(lr),1)
      return
 42   continue
C     ------------------------
C     special case with Updates 
C     [w,xu]=corr('updt',x1,[y1],w0) etc....
      ichaud=0
      mode=0
c     -------------x1 value 
      top2=top2+1
      if (.not.getrmat(fname,topk,top2,mx,nx,lx)) return
      mnx=mx*nx
c     --------------y1 value or w0
      top2=top2+1
      if (.not.getmat(fname,topk,top2,it,my,ny,ly,lyc)) return 
      mny=ny*my
      if(mny.eq.mnx) then
         mode=1
         if ( it.eq.1) then 
            buf= fname // ' y1 should be a real matrix'
            call error(999)
            return
         endif
      endif
c     --------------third or second argument 
      if(mode.eq.1) top2=top2+1
      if (.not.getmat(fname,topk,top2,it,mz,nz,lzr,lzi)) return 
      mfft=mz*nz
C     if w0 is real we add space for storing an imag part 
      if (it.eq.0) then
         if (.not.cremat(fname,topw,0,1,mfft,lzi,lc)) return
         topw=topw+1
         call dset(mfft,0.d0,stk(lzi),1)
      endif
      if(mode.eq.1.and.rhs.eq.5) ichaud=1
      if(mode.eq.0.and.rhs.eq.4) ichaud=1
c     -----------last argument + working areas 
      if (.not.cremat(fname,topw,1,1,mfft,lwr,lwi)) return
      call dset(2*mfft,0.d0,stk(lwr),1)
      if(ichaud.eq.1) then
         top2=top2+1
         if (.not.getmat(fname,topk,top2,it,mx0,nx0,
     $        lxr,lxi))   return 
         nbx=mx0*nx0
         if ( nbx.gt.mfft) then 
            buf = fname // 'last argument has wrong size '
            call error(999)
            return
         endif
         call unsfdcopy(nbx,stk(lxr),1,stk(lwr),1)
      endif
C     ---------calling the function 
      if(mode.eq.1) then
         call cmpse3(mfft,mnx,mode,stk(lx),stk(ly),stk(lwr),stk(lwi),
     1        stk(lzr),stk(lzi),ierr,ichaud,nbx)
      endif
      if(mode.eq.0) then
         call cmpse3(mfft,mnx,mode,stk(lx),yyy,stk(lwr),stk(lwi),
     1        stk(lzr),stk(lzi),ierr,ichaud,nbx)
      endif
C     ---------back to Scilab
      if(ierr.gt.0) then
         buf='fft call: needs a power of 2'
         call error(999)
         return
      endif
      if (lhs.eq.2) then 
C        keep this in a safe place before the copy for lhs1
         call unsfdcopy(nbx,stk(lx+mnx-nbx),1,stk(lwr),1)
      endif
      if(lhs.ge.1) then
         top=topk-rhs+1
         if (.not.cremat(fname,top,1,mz,nz,lres,lresi)) return
         call unsfdcopy(mfft,stk(lzr),1,stk(lres),1)
         call unsfdcopy(mfft,stk(lzi),1,stk(lresi),1)
      endif
      if(lhs.eq.2) then
         top=top+1
         if (.not.cremat(fname,top,0,1,nbx,lres,lresi)) return
         call unsfdcopy(nbx,stk(lwr),1,stk(lres),1)         
      endif
      return
      end
c     =======================================

File: scilab/modules/signal_processing/sci_gateway/c/sci_corr.c
/*
 * Scilab ( http://www.scilab.org/ ) - This file is part of Scilab
 * Copyright (C) 2006 - INRIA - Allan CORNET
 * Copyright (C) 2009 - Digiteo - Vincent LIARD
 *
 * This file must be used under the terms of the CeCILL.
 * This source file is licensed as described in the file COPYING, which
 * you should have received as part of this distribution.  The terms
 * are also available at
 * http://www.cecill.info/licences/Licence_CeCILL_V2-en.txt
 */

#include "api_scilab.h"
#include "MALLOC.h"
#include "gw_signal.h"
#include "stack-c.h"
#include "Scierror.h"
#include "localization.h"
#include "../../../statistics/src/c/sci_string_matrix.h"

extern int C2F(scicorr)(char *id,unsigned long fname_len );

int corr_default(char *fname, int* _piKey);
int corr_updates(char *fname, int* _piKey);
int corr_fft(char *fname, int* _piKey);

/* dispatch to specialized gateways according to mode (default, 'updt', 'fft') */
int sci_corr(char *fname, int* _piKey)
{
  char *name;
  int type;
  int *arg;

  CheckRhs(2,5);
  CheckLhs(1,2);
  getVarAddressFromPosition(_piKey, 1, &arg);
  getVarType(_piKey, arg, &type);
	if (type == sci_strings) {
    name = create_string(_piKey, 1);
    switch (name[0]) {
    case 'f': /* fft */
      CheckRhs(4,5);
      corr_fft(fname, _piKey);
      break;
    case 'u': /* updates */
      CheckRhs(3,5);
      corr_updates(fname, _piKey);
      break;
    default:
      Scierror(999, _("%s: Wrong value for input argument #%d: Must be in the set {%s}.\n"), fname, 1, "'fft', 'update'");
      return 1;
    }
    destroy_string(name);
  }
  else { /* default */
    CheckRhs(2,3);
    corr_default(fname, _piKey);
  }
  return 0;
}

int corr_default(char *fname, int* _piKey) {

  extern void C2F(tscccf)(double *x, double *y, int *length,
			  double *cxy, double *xymean, int *lag, int *error);

  int length, lags_number, error, rows1, cols1, rows2, cols2, rows3, cols3;
  double *x, *y, *xvar /* crossvariance */, *mean, *data;
  int *arg;
  int type;

  getVarAddressFromPosition(_piKey, 1, &arg);
  getVarDimension(_piKey, arg, &rows1, &cols1);
  length = rows1 * cols1;
  getVarType(_piKey, arg, &type);
  if ((type != sci_matrix) ||
      (rows1 != 1 && cols1 != 1)) {
    Scierror(999, _("%s: Wrong size for input argument #%d: A vector expected.\n"), fname, 1);
    return 1;
  }
  getMatrixOfDouble(_piKey, arg, &rows1, &cols1, &x);

  /* interpret 2nd argument according to number of arguments */
  getVarAddressFromPosition(_piKey, 2, &arg);
  getMatrixOfDouble(_piKey, arg, &rows2, &cols2, &data);
  if (Rhs == 2) {
    if (rows2 != 1 || cols2 != 1) {
      Scierror(999, _("%s: Wrong size for input argument #%d: A scalar expected.\n"), fname, 2);
      return 1;
    }
    y = x;
    lags_number = (int)data[0];
  }
  else {
    if (rows1 != rows2 || cols1 != cols2) {
      Scierror(999, _("%s: Incompatible input arguments #%d and #%d': Same sizes expected.\n"), fname, 1, 2);
      return 1;
    }
    y = data;
    getVarAddressFromPosition(_piKey, 3, &arg);
    getMatrixOfDouble(_piKey, arg, &rows3, &cols3, &data);
    lags_number = (int)data[0];
  }
  allocMatrixOfDouble(_piKey, Rhs + 1, 1, lags_number, &xvar);
  allocMatrixOfDouble(_piKey, Rhs + 2, 1, (x == y) ? 1 : 2, &mean);
  C2F(tscccf)(x, y, &length, xvar, mean, &lags_number, &error);
  LhsVar(1) = Rhs + 1;
  LhsVar(2) = Rhs + 2;
  PutLhsVar();
  return error;
}

int corr_fft(char *fname, int* _piKey) {

  extern void C2F(cmpse2)(int *m, int *n, int *mode,
			  void (*bgetx)(double *, int *, int *),
			  void (*bgety)(double *, int *, int *),
			  double *xa, double *xr, double *xi,
			  double *zr, double *zi, int *error);
  extern void C2F(bgetx)(double *x, int *increment, int *start_idx);
  extern void C2F(bgety)(double *x, int *increment, int *start_idx);
  extern void C2F(setdgetx)(char *name, int *rep);
  extern void C2F(setdgety)(char *name, int *rep);

  enum {self = 2, cross = 3} mode = self;
  int error, rows, cols, n, mm, lag;
  double *r;
  double *xa, *xr, *xi, *zr, *zi, *data;
  int return2_length, one = 1;
  int *arg;

  /* taken from DllmainSignal_processing.c */
  /* the very f*** global referred to underneath */
  extern struct {
    int kgxtop, kgytop, ksec, kisc;
  } C2F(corradr);

  getVarAddressFromPosition(_piKey, Rhs, &arg);
  getMatrixOfDouble(_piKey, arg, &rows, &cols, &data);
  lag = (int)(*data);
  mm = 2 * lag;
  getVarAddressFromPosition(_piKey, Rhs - 1, &arg);
  getMatrixOfDouble(_piKey, arg, &rows, &cols, &data);
  n = (int)(*data);
  mode = (Rhs == 4 ? self : cross);

  /** @todo add code to manage calls to compiled functions */

  /* setting the f*** (!@&£*% etc) global parameter for cmpse2 */
  C2F(corradr).kgxtop = Top - Rhs + 2;
  C2F(corradr).kgytop = Top - Rhs + 3; /* useless for self correlation but doesn't hurt */
  C2F(corradr).ksec = Top;
  C2F(corradr).kisc = Top + 1;

  xa = (double *)MALLOC(mm * sizeof(double));
  xr = (double *)MALLOC(mm * sizeof(double));
  xi = (double *)MALLOC(mm * sizeof(double));
  zr = (double *)MALLOC(mm * sizeof(double));
  zi = (double *)MALLOC(mm * sizeof(double));
  {
    int imode = mode; /* enum* not equiv to int* :'( */
    C2F(cmpse2)(&mm, &n, &imode,
		C2F(bgetx), C2F(bgety),
		xa, xr, xi, zr, zi, &error);
  }
  allocMatrixOfDouble(_piKey, Rhs + 1, 1, lag, &r);
  C2F(dcopy)(&lag, xa, &one, r, &one);

  LhsVar(1) = Rhs + 1;
  return2_length = Rhs - 3;
  allocMatrixOfDouble(_piKey, Rhs + 2, 1, return2_length, &r);
  C2F(dcopy)(&return2_length, xr, &one, r, &one);
  FREE(zi);
  FREE(zr);
  FREE(xi);
  FREE(xr);
  FREE(xa);
  LhsVar(2) = Rhs + 2;
  PutLhsVar();
  return 0;
}

/* assert: RhsVar must have been checked to hold 4 to 5 elements */
int corr_updates(char *fname, int* _piKey) {

  extern void C2F(cmpse3)(int *m, int *n, int *mode, double *x, double *y,
			  double *xr, double *xi,
			  double *zr, double *zi,
			  int *error, int *ichaud, int *nbx);

  enum {self = 0, cross = 1} mode = self;
  int update = 0;
  int mfft, wpos;
  int length, error, rows, cols;
  double *x, *y,
    *zr, *zi, *wr, *wi, *x_real,
    *rr, *ri, y_value;
  int *arg;
  int one = 1, nbx = 0;
  double zero = 0L;

  /* x */
  getVarAddressFromPosition(_piKey, 2, &arg);
  getMatrixOfDouble(_piKey, arg, &rows, &cols, &x);
  length = rows * cols;
  getVarAddressFromPosition(_piKey, 3, &arg);
  getVarDimension(_piKey, arg, &rows, &cols);
  if (length == rows * cols) {
    mode = cross;
  }
  if ((mode == cross && Rhs == 5) ||
      (mode == self && Rhs == 4)) {
    update = 1;
  }

  /* y */
  if (mode == cross) {
    CheckRhs(4,5);
    if (isVarComplex(_piKey, arg)) {
      Scierror(999, _("%s: Input argument #%d must be real.\n"), fname, 3);
      return 1;
    }
    getMatrixOfDouble(_piKey, arg, &rows, &cols, &y);
  }
  else {
    y_value = 0;
    y = &y_value;
  }

  /* w */
  wpos = (mode == self) ? 3 : 4;
  /* dimension ought to be a power of 2 (could be checked here) */
  /* GetVarDimension(wpos, &rows, &cols); */
  getVarAddressFromPosition(_piKey, wpos, &arg);
  if (isVarComplex(_piKey, arg)) {
    getComplexMatrixOfDouble(_piKey, arg, &rows, &cols, &zr, &zi);
    mfft = rows * cols;
  }
  else {
    getMatrixOfDouble(_piKey, arg, &rows, &cols, &zr);
    mfft = rows * cols;
    zi = (double *)MALLOC(mfft * sizeof(double));
    C2F(dset)(&mfft, &zero, zi, &one);
  }

  /* xu */
  wr = (double *)MALLOC(mfft * sizeof(double));
  wi = (double *)MALLOC(mfft * sizeof(double));
  C2F(dset)(&mfft, &zero, wr, &one);
  C2F(dset)(&mfft, &zero, wi, &one);
  if (update) { /* if xu provided, update it */
    getVarAddressFromPosition(_piKey, wpos + 1, &arg);
    getMatrixOfDouble(_piKey, arg, &rows, &cols, &x_real);
    nbx = rows * cols;
    if (nbx > mfft) {
      /* TODO: make error msg more accurate (2nd dim >= dim(arg)) */
      Scierror(999, _("%s: Wrong size for input argument #%d: A %d-by-%d matrix expected.\n"), fname, 5, 1, 999);
      return 1;
    }
    C2F(dcopy)(&nbx, x_real, &one, wr, &one);
  }
  else {
    nbx = 0;
  }

  C2F(cmpse3)(&mfft, &length, (int *) &mode, x, y,
	      wr, wi, zr, zi,
	      &error, (int *) &update, &nbx);

  if (error) {
    Scierror(999, _("%s: Wrong size for input argument #%d: Should be a power of 2"), fname, wpos);
    return error;
  }
  /** @warning x built of data on the stack that's reaffected by matrix allocation
     -> copied to keep it safe for lhs2 output */
  C2F(dcopy)(&nbx, &(x[length-nbx]), &one, wr, &one);
  allocComplexMatrixOfDouble(_piKey, Rhs + 1, 1, mfft, &rr, &ri);
  C2F(dcopy)(&mfft, zr, &one, rr, &one);
  C2F(dcopy)(&mfft, zi, &one, ri, &one);
  LhsVar(1) = Rhs + 1;
  if (Lhs == 2) {
    allocMatrixOfDouble(_piKey, Rhs + 2, 1, nbx, &rr);
    C2F(dcopy)(&nbx, wr, &one, rr, &one);
    LhsVar(2) = Rhs + 2;
  }
  PutLhsVar();
  FREE(wr);
  FREE(wi);
  return error;
}
